lsp/nls/src/position.rs

```diff
@@ -0,0 +1,189 @@
+                all_ranges.push((
+                    Range {
+                        start: pos.start.0,
+                        end: pos.end.0,
+                    },
+                    term.clone(),
+                ));
+            }
+            TraverseControl::<()>::Continue
+        });
+
+        // We rely on the invariant that if two ranges overlap then one is contained in the
+        // other. That is, the ranges form a tree. This sort order corresponds to pre-order
+        // traversal of that tree.
+        //
+        // (It would be nice if we could build the lookup table directly from the traversal
+        // in traverse_ref, but (1) the state tracking is fiddly and (2) the traversal order
+        // in traverse_ref can be weird; for example, a `Term::Annotated` sometimes contains
+        // children whose position is outside the position of the parent.)
+        all_ranges.sort_by_key(|(range, _term)| (range.start, std::cmp::Reverse(range.end)));
+        let mut all_ranges = all_ranges.into_iter().peekable();
+
+        // Now we iterate over the tree of ranges to make a disjoint set of ranges.
+        // `stack` is the path in the tree leading to the node that we are currently visiting.
+        let mut stack = Vec::new();
+        let mut next = all_ranges.next();
+
+        // We accumulate ranges using this closure, which guarantees that they are non-overlapping.
+        // (This is important: RangeMap panics if they are not.)
+        let mut disjoint = Vec::new();
+        // The last position we've added to `disjoint`. Gets bumped automatically every time we
+        // push a new range.
+        let mut pos = next
+            .as_ref()
+            .map(|(range, _term)| range.start)
+            .unwrap_or_default();
+        let mut push_range = |pos: &mut u32, end: u32, term: RichTerm| {
+            debug_assert!(*pos <= end);
+            if *pos < end {
+                disjoint.push((Range { start: *pos, end }, RichTermPtr(term)));
+                *pos = end;
+            }
+        };
+
+        while let Some((cur, term)) = next {
+            // If the next interval overlaps us, then we must contain it and it is our child.
+            // Otherwise, we are a leaf and it is a sibling or a cousin or something.
+            let next_start = all_ranges.peek().map(|(r, _term)| r.start);
+            match next_start {
+                Some(next_start) if cur.end > next_start => {
+                    // It is our child.
+                    push_range(&mut pos, next_start, term.clone());
+                    stack.push((cur, term));
+                    next = all_ranges.next();
+                }
+                _ => {
+                    // It is a sibling/cousin
+                    push_range(&mut pos, cur.end, term.clone());
+                    next = stack.pop().or_else(|| all_ranges.next());
+                }
+            }
+        }
```

Comment:
I must say I'm still having a bit of an issue understanding why we do all of that. I think we agreed that intervals must be non overlapping. If we want to check that, I guess we can simply iterate over the sorted position arrays in debug mode, like 

```rust
debug_assert!({
 ...Check that intervals aren't overlapping
})
```

Otherwise, if intervals are assumed to be non overlapping, can't we just sort the whole thing once and call it a day? Then we can binary search into it. I believe this is currently exactly what we do already in the linearization implementation. Is there any reason to do the same here? What's the point of building tree path, and poping from the stack, revisiting maybe some parent nodes several times, etc. ?

Reply:
The initial set of intervals is not non-overlapping: parents (usually) contain children. This isn't the end of the world (it's what the current linearizer does) but it means that lookup can involve a scan that's linear in the tree height.

This extra traversal makes the intervals disjoint. For example, if a parent is [0, 10) and its children are [1, 2) and [5, 6) then it first visits the parent and adds the interval [0, 1), then it visits the first child and adds the interval [1, 2), then it returns to the parent to add [2, 5), then visits the second child, and so on.

I was originally trying to do this from within `traverse_ref` but the non-nestedness of the positions made that fail. Actually, even though this approach effectively does an extra traversal it has some other advantages (less allocation, for a start) so I'm not sure how the speed compares.

Reply:
Ok, I think my confusion came  from that I probably had the wrong mental definition of overlapping. I thought of overlapping as "two intervals that intersect but where none is contained in the other", so I didn't understand why you needed to make ranges non-overlapping - they should already be this way. But you really mean "two intervals that intersect". It's clearer now, thanks!

Still, it wasn't obvious at all to me at the first read, as to what this transformation was all about. I think it could benefit from being pulled in some helper function, and be well documented, for example with the example you just gave on `[0, 10)`.